File: threefive/upid.py
"""
threefive/upid.py

threefve.upid exposes one function,
upid_decoder for use by the
SegmentationDescriptor class.

"""


def upid_decoder(bitbin, upid_type, upid_length):
    """
    upid_decoder
    decodes segmentation_upids by type,
    from a bitbin instance.

    Used by the SegmentationDescriptor class.
    """
    upid_map = {
        0x01: ["Deprecated", _decode_uri],
        0x02: ["Deprecated", _decode_uri],
        0x03: ["AdID", _decode_uri],
        0x04: ["UMID", _decode_umid],
        0x05: ["ISAN", _decode_isan],
        0x06: ["ISAN", _decode_isan],
        0x07: ["TID", _decode_uri],
        0x08: ["AiringID", _decode_air_id],
        0x09: ["ADI", _decode_uri],
        0x0A: ["EIDR", _decode_eidr],
        0x0B: ["ATSC", _decode_atsc],
        0x0C: ["MPU", _decode_mpu],
        0x0D: ["MID", _decode_mid],
        0x0E: ["ADS Info", _decode_uri],
        0x0F: ["URI", _decode_uri],
        0x10: ["UUID", _decode_uri],
        0xFD: ["Unknown", _decode_uri],
    }
    if upid_type not in upid_map.keys():
        upid_type = 0xFD
    return upid_map[upid_type][0], upid_map[upid_type][1](bitbin, upid_length)


def upid_encoder(nbin, upid_type, upid_length, seg_upid):
    """
    upid_encoder
    encodes segmentation_upid values to bites
    added to the nbin instance.

    Used by the SegmentationDescriptor class.
    """

    upid_map = {
        0x02: ["Deprecated", _encode_uri],
        0x03: ["AdID", _encode_uri],
        0x04: ["UMID", _encode_umid],
        0x07: ["TID", _encode_uri],
        0x0B: ["ATSC", _encode_atsc],
        0x09: ["ADI", _encode_uri],
        0x0A: ["EIDR", _encode_eidr],
        0x0C: ["MPU", _encode_mpu],
        0x0D: ["MID", _encode_mid],
        0x0E: ["ADS Info", _encode_uri],
        0x0F: ["URI", _encode_uri],
        0x10: ["UUID", _encode_uri],
    }

    upid_map_too = {
        0x05: ["ISAN", _encode_isan],
        0x06: ["ISAN", _encode_isan],
        0x08: ["AiringID", _encode_air_id],
    }

    if upid_type in upid_map:
        upid_map[upid_type][1](nbin, seg_upid)

    if upid_type in upid_map_too:
        upid_map_too[upid_type][1](nbin, seg_upid, upid_length)


def _decode_air_id(bitbin, upid_length):
    return bitbin.as_hex(upid_length << 3)


def _encode_air_id(nbin, seg_upid, upid_length):
    nbin.add_hex(seg_upid, (upid_length << 3))


def _decode_atsc(bitbin, upid_length):
    return {
        "TSID": bitbin.as_int(16),
        "reserved": bitbin.as_int(2),
        "end_of_day": bitbin.as_int(5),
        "unique_for": bitbin.as_int(9),
        "content_id": bitbin.as_ascii(((upid_length - 4) << 3)),
    }


def _encode_atsc(nbin, seg_upid):
    nbin.add_int(seg_upid["TSID"], 16)
    nbin.add_int(seg_upid["reserved"], 2)
    nbin.add_int(seg_upid["end_of_day"], 5)
    nbin.add_int(seg_upid["unique_for"], 9)
    nbin.add_bites(seg_upid["content_id"].encode("utf-8"))


def _decode_eidr(bitbin, upid_length):
    if upid_length < 12:
        raise Exception(f"upid_length is {upid_length} should be 12 bytes")
    pre = bitbin.as_int(16)
    post = []
    bit_count = 80
    while bit_count:
        bit_count -= 16
        post.append(bitbin.as_hex(16)[2:])
    return f"10.{pre}/{'-'.join(post)}"


def _encode_eidr(nbin, seg_upid):
    pre, post = seg_upid[3:].split("/", 1)
    nbin.add_int(int(pre), 16)
    nbin.add_hex(post.replace("-", ""), 80)


def _decode_isan(bitbin, upid_length):
    return bitbin.as_hex(upid_length << 3)


def _encode_isan(nbin, seg_upid, upid_length):
    nbin.add_hex(seg_upid, (upid_length << 3))


def _decode_mid(bitbin, upid_length):
    upids = []
    ulb = upid_length << 3
    while ulb:
        upid_type = bitbin.as_int(8)  # 1 byte
        ulb -= 8
        upid_length = bitbin.as_int(8)
        ulb -= 8
        upid_type_name, segmentation_upid = upid_decoder(bitbin, upid_type, upid_length)
        mid_upid = {
            "upid_type": upid_type,
            "upid_type_name": upid_type_name,
            "upid_length": upid_length,
            "segmentation_upid": segmentation_upid,
        }
        ulb -= upid_length << 3
        upids.append(mid_upid)
    return upids


def _encode_mid(nbin, seg_upid):
    for mid_upid in seg_upid:
        nbin.add_int(mid_upid["upid_type"], 8)
        nbin.add_int(mid_upid["upid_length"], 8)
        upid_encoder(
            nbin,
            mid_upid["upid_type"],
            mid_upid["upid_length"],
            mid_upid["segmentation_upid"],
        )


def _decode_mpu(bitbin, upid_length):
    ulbits = upid_length << 3
    mpu_data = {
        "format_identifier": bitbin.as_int(32),
        "private_data": bitbin.as_raw(ulbits - 32),
    }
    return mpu_data


def _encode_mpu(nbin, seg_upid):
    nbin.add_int(seg_upid["format_identifier"], 32)
    nbin.add_bites(seg_upid["private_data"])


def _decode_umid(bitbin, upid_length):
    chunks = []
    ulb = upid_length << 3
    while ulb:
        chunks.append(bitbin.as_hex(32).split("x", 1)[1])
        ulb -= 32
    return ".".join(chunks)


def _encode_umid(nbin, seg_upid):
    chunks = seg_upid.split(".")
    for chunk in chunks:
        nbin.add_hex(chunk, 32)


def _decode_uri(bitbin, upid_length):
    if upid_length > 0:
        return bitbin.as_ascii(upid_length << 3)
    return 0


def _encode_uri(nbin, seg_upid):
    seg_upid = seg_upid.encode("utf-8")
    nbin.add_bites(seg_upid)

File: threefive/upid.go
package cuei

import (
	"fmt"
)

// Upid is the Struct for Segmentation Upida
type Upid struct {
	Name             string `json:",omitempty"`
	UpidType         uint8  `json:",omitempty"`
	Value            string `json:",omitempty"`
	TSID             uint16 `json:",omitempty"`
	Reserved         uint8  `json:",omitempty"`
	EndOfDay         uint8  `json:",omitempty"`
	UniqueFor        uint16 `json:",omitempty"`
	ContentID        string `json:",omitempty"`
	Upids            []Upid `json:",omitempty"`
	FormatIdentifier string `json:",omitempty"`
	PrivateData      string `json:",omitempty"`
}

// UpidDecoder calls a method based on upidType
func (upid *Upid) Decoder(gob *Gob, upidType uint8, upidlen uint8) {

	upid.UpidType = upidType

	switch upidType {
	case 0x01, 0x02:
		upid.Name = "Deprecated"
		upid.URI(gob, upidlen)
	case 0x03:
		upid.Name = "AdID"
		upid.URI(gob, upidlen)
	case 0x05, 0x06:
		upid.Name = "ISAN"
		upid.ISAN(gob, upidlen)
	case 0x07:
		upid.Name = "TID"
		upid.URI(gob, upidlen)
	case 0x08:
		upid.Name = "AiringID"
		upid.AirID(gob, upidlen)
	case 0x09:
		upid.Name = "ADI"
		upid.URI(gob, upidlen)
	case 0x0a:
		upid.Name = "EIDR"
		upid.EIDR(gob, upidlen)
	case 0x0b:
		upid.Name = "ATSC"
		upid.ATSC(gob, upidlen)
	case 0x0c:
		upid.Name = "MPU"
		upid.MPU(gob, upidlen)
	case 0x0d:
		upid.Name = "MID"
		upid.MID(gob, upidlen)
	case 0x0e:
		upid.Name = "ADS Info"
		upid.URI(gob, upidlen)
	case 0x0f:
		upid.Name = "URI"
		upid.URI(gob, upidlen)
	case 0x10:
		upid.Name = "UUID"
		upid.URI(gob, upidlen)
	default:
		upid.Name = "UPID"
		upid.URI(gob, upidlen)
	}
}

// Decode for AirId
func (upid *Upid) AirID(gob *Gob, upidlen uint8) {
	upid.Value = gob.Hex(uint(upidlen << 3))
}

// Decode for Isan Upid
func (upid *Upid) ISAN(gob *Gob, upidlen uint8) {
	upid.Value = gob.Ascii(uint(upidlen << 3))
}

// Decode for URI Upid
func (upid *Upid) URI(gob *Gob, upidlen uint8) {
	upid.Value = gob.Ascii(uint(upidlen) << 3)
}

// Decode for ATSC Upid
func (upid *Upid) ATSC(gob *Gob, upidlen uint8) {
	upid.TSID = gob.UInt16(16)
	upid.Reserved = gob.UInt8(2)
	upid.EndOfDay = gob.UInt8(5)
	upid.UniqueFor = gob.UInt16(9)
	upid.ContentID = gob.Ascii(uint((upidlen - 4) << 3))
}

// Decode for EIDR Upid
func (upid *Upid) EIDR(gob *Gob, upidlen uint8) {
	if upidlen == 12 {
		head := gob.UInt64(16)
		tail := gob.Hex(80)
		upid.Value = fmt.Sprintf("10%v/%v", head, tail)
	}
}

// Decode for MPU Upid
func (upid *Upid) MPU(gob *Gob, upidlen uint8) {
	ulb := uint(upidlen) << 3
	upid.FormatIdentifier = gob.Hex(32)
	upid.PrivateData = gob.Ascii(ulb - 32)
}

// Decode for MID Upid
func (upid *Upid) MID(gob *Gob, upidlen uint8) {
	var i uint8
	i = 0
	for i < upidlen {
		utype := gob.UInt8(8)
		i++
		ulen := gob.UInt8(8)
		i++
		i += ulen
		var mupid Upid
		mupid.Decoder(gob, utype, ulen)
		upid.Upids = append(upid.Upids, mupid)
	}
}
